src/move.c

```diff
@@ -3161,33 +3145,39 @@ pagescroll(int dir, long count, int half)
 	if (count)
 	    curwin->w_p_scr = MIN(curwin->w_height, count);
 	count = MIN(curwin->w_height, curwin->w_p_scr);
+	int curscount = count;
 
-	// Don't scroll if we already know that it will reveal end of buffer lines.
-	if (dir == BACKWARD
-	    || (curwin->w_botline - 1 < buflen)
-	    || (curwin->w_p_sms && curwin->w_botline - 1 == buflen
-		&& curwin->w_skipcol < linetabsize(curwin, buflen)))
+	// Adjust count so as to not reveal end of buffer lines.
+	if (dir == FORWARD)
 	{
-	    nochange = scroll_with_sms(dir, &count);
-	    validate_botline();
-	    // Hide any potentially revealed end of buffer lines.
-	    if (!nochange && curwin->w_botline - 1 == buflen)
+	    int n = plines_win(curwin, curwin->w_topline, FALSE);
+		n -= adjust_plines_for_skipcol(curwin);
+	    if (n - count < curwin->w_height && curwin->w_topline < buflen)
+		n += plines_m_win(curwin, curwin->w_topline + 1, buflen, TRUE);
+	    if (n - count < curwin->w_height)
+		count = n - curwin->w_height;
+	}
+
+	if (count > 0)
+	{
+	    validate_cursor();
+	    int	prev_wrow = curwin->w_wrow;
+	    nochange = scroll_with_sms(dir, count);
+	    if (!nochange)
 	    {
-		curwin->w_cursor.lnum = buflen;
-		scroll_cursor_bot(0, TRUE);
+		validate_cursor();
```

Comment:
Instead of messing with curswant here which is already set accordingly in scrollup/down(), validate the cursor and bring it to the row it was previously on. I think the cursor position is much more predictable now with 'nostartofline'. 

There are still some bugs with 'scrolloff' with large lines in a small window that prevent scrolling or corrupt the screen. Those were also present before patch 9.1.0211, but now also apply to (half)-page scrolling in addition to Ctrl-E/Y.